File: python/beginners_selection/abc049c.py
# https://atcoder.jp/contests/abs/tasks/arc065_a

s = input()
l = ["dream", "dreamer", "erase", "eraser"]
t = s
while len(t) > 0:
    for word in l:
        if t.endswith(word):
            index = len(t) - len(word)
            if index == 0:
                print("YES")
                exit()
            elif index > 0:
                t = t[:index]
                break
        if word == l[len(l) - 1]:
            print("NO")
            exit()

print("NO")

File: cpp/beginners_selection/abc049c.cpp
// https://atcoder.jp/contests/abs/tasks/arc065_a

#include<iostream>
using namespace std;

bool ends_with(const string& s, const string& suffix) {
   if (s.size() < suffix.size()) return false;
   return equal(rbegin(suffix), rend(suffix), rbegin(s));
}

int main() {
    string s;
    cin >> s;

    string words[4] = {"dream", "dreamer", "erase", "eraser"};
    bool found = false;
    int i = 0;
    while (i < 4 && !found) {
        string word = words[i];
        if (ends_with(s, word)) {
            s = s.erase(s.size() - word.size());
            if (s.empty()) {
                found = true;
            } else {
                i = 0;
            }
        } else {
            i++;
        }
    }

    string res = found ? "YES" : "NO";
    cout << res << endl;
    return 0;
}
